### ISSUE TITLE
importing scipy.io and matplotlib.pyplot in InterfaceWithPython hangs module

### ISSUE THREAD

**Issue by jessdtate** [SCIInstitute/SCIRun#1570]
### Description

Trying to import the scipy.io or matplotlib.pyplot in the IWP module or in a script called by IWP will hang the module (test_scripting_IV.srn5).  These same functions work fine if they are called from the python console.  

[test_scripting_IV.srn5.zip](https://github.com/SCIInstitute/SCIRun/files/586816/test_scripting_IV.srn5.zip)

### Steps to Reproduce

1. open IWP and put import scipy.io or import matplotlib.pyplot in the text field.
2. execute module
Module will hang and SCIRun will need to be killed.
3. if the previous lines are put in the python console, no problems.  
4. load test_scripting_IV.srn5, and execute
module will hang as before.
5. copy the text from test_scripting_IV.srn5 IWP to python console.  
code will "work" with a bug as shown in #1537 

**Expected behavior:** 

it should work the same in the IWP as in the python console

**Actual behavior:**

hangs the module and scirun

### Versions

os x 10.10 and 10.8
scirun v5.0-beta.E-96-g17438de

**Comment by jessdtate** [comment]
Here is another line that hangs the iwp module but not the python console: np.finfo(float).eps.  Use this code to test:

import numpy as np
print(np.finfo(float).eps)

**Comment by jessdtate** [comment]
Another set of code that hangs the IWP but works in the python console.  This one seems to have problems with return an error/warning and an inf value.  Similar functions that I have tried return inf and an error/warning have the same issue (np.linalg.norm, for example)

import numpy as np
g = np.identity(9)
c1 = np.linalg.cond(g)
print(c1)
g[0,:] =0
c2 = np.linalg.cond(g)
print(c2)

The first one (c1) works because the result is a finite value.

**Comment by jessdtate** [comment]
in the last example (np.linalg.cond) the terminal does print the warning and the result after a while.  It takes several minutes when it only takes a second in the python console.

**Comment by dcwhite** [comment]
@jessdtate I was finally able to import numpy on my new Mac running 10.12. I never got pip to work, so I built numpy from source using the latest code (along with building the latest cython, downloading latest python installtools, and homebrewing latest gfortran compiler). This worked all within SCIRun's custom python install. Here are my findings:

1. The simple `print(np.finfo(float).eps)` line did not hang in either console or module.
2. The infinite condition number code did replicate the hanging-in-module issue, so I will continue to investigate that.
3. I have not built scipy yet, stay tuned.

I am considering providing a script that will download and build all these libraries after the user builds SCIRun. That will be made into a separate issue after these strange behaviors are resolved.

**Comment by jessdtate** [comment]
it only hangs in the IWP for me.

**Comment by dcwhite** [comment]
Do you remember how long the infinite cond output took with IWP? Mine is still hung after 30+ minutes.

**Comment by jessdtate** [comment]
I never got it to finish.  I've waited hours for it.

**Comment by dcwhite** [comment]
Oh, above it said "several minutes."

**Comment by jessdtate** [comment]
oh, on that example I don't remember, but I think it was around there.   I will run it again.  The example in the fwd/inv toolkit I definitely ran a long time.

**Comment by jessdtate** [comment]
I retried it and it ran over the weekend without finishing.

**Comment by dcwhite** [comment]
Could you try running it through the debugger?

**Comment by dcwhite** [comment]
Beta for sure

**Comment by dcwhite** [comment]
@jessdtate This is the main one impacting UQ now right?

**Comment by jessdtate** [comment]
scipy is problematic (more than scipy.io) and is holding up the UQ.  Matplotlib is a not as crucial, those it would be helpful for the UQ.

**Comment by jessdtate** [comment]
`print(np.finfo(float).eps)` may be dependent on context.  If I run it with just the import statement before, it hangs. With other code it seems to work.

**Comment by jessdtate** [comment]
`import scipy.io` seems to still be problematic in later numpy versions

**Comment by dcwhite** [comment]
**Investigation findings (2026-06-12)**

The namespace scoping fix in #1569 (passing globals as locals in run_script) will fix the import not being found. But the *hang* on scipy.io and matplotlib is a separate, deeper issue.

**Root cause of the hang:** scipy.io and matplotlib C extensions perform thread-unsafe GUI or OS-level operations during import. When imported on the IWP execution thread (not the main thread), they attempt to acquire locks or interact with the OS event loop that are already held by the main thread, causing a deadlock.

Specific evidence from comments: np.linalg.cond on a singular matrix produces a numpy warning, which triggers numpy's warning machinery. This warning dispatch path apparently tries to acquire a lock that the main thread is also waiting on — explaining the infinite hang only on singular/infinite values.

**Fix approaches:**
1. Run IWP Python execution on the main thread (or dispatch warnings back to main thread) — invasive
2. Suppress numpy warnings in IWP by default: import warnings; warnings.filterwarnings('ignore') at module init
3. For scipy.io specifically: ensure the import happens on the main thread via a queued signal before execution begins
4. Redirect stderr/stdout through a thread-safe queue during execution (partial mitigation)

This is also blocked by #1295 (pip/SSL) since scipy needs pip to install and test.

**Effort:** Medium-high. The warning-dispatch deadlock is the most likely culprit and may be fixable by redirecting Python warnings to a thread-safe handler.
